### ISSUE TITLE
Build does not populate properties that have customizations elsewhere

### ISSUE THREAD

**Issue by drewrandall** [AutoFixture/AutoFixture#772]
Using Build along with customizations results in a an unset property on the built object. In this example, I would expect either "Cat" or Guid.NewGuid(), but not null.

Observe:

```
[TestClass]
public class AutoFixtureTest
{
    public class TestObject
    {
        public string A { get; set; }
        public string B { get; set; }
        public string C { get; set; }
    }

    [TestMethod]
    public void ShouldPopulateABC()
    {
        var fixture = new Fixture();
        fixture.Customize<TestObject>(x => x.With(p => p.A, "Dog"));
        fixture.Customize<TestObject>(x => x.With(p => p.C, "Cat"));

        var testObject = fixture.Build<TestObject>().With(x => x.A, "Zebra").Create();

        Assert.AreEqual("Zebra", testObject.A, "A is Zebra.");
        Assert.IsNotNull(testObject.B, "B is null.");
        Assert.IsNotNull(testObject.C, "C is null.");
    }
}
```
Results in: 
```
Assert.IsNotNull failed. C is null.
```

**Comment by drewrandall** [comment]
Forgot to mention this is version 3.50.2

**Comment by zvirja** [comment]
@drewrandall Please see the [similar]https://stackoverflow.com/questions/38688932/how-to-use-autofixture-to-build-with-customized-properties-while-keeping-type-cu) question the stack overflow - it should cover your case.

Let me know if your case differs 😉

**Comment by drewrandall** [comment]
@zvirja thanks for the link. It certainly is related. I am already convinced that customizations are superior to build. This actually came about as I try to show my team.

I'm still confused by why B and C perform differently in my example, and I think there is a real bug here.

The build method documentation says: 
> Customizes the creation algorithm for a single object, effectively turning off all Customizations on the Ploeh.AutoFixture.IFixture.

My expectation is that the customizations specified outside of Build are _irrelevant_, but their existence is causing B and C to be handled differently. At the least, they should be handled the same way. Ideally, they should get the default behavior, which I understand to be a random string created by the fixture.

**Comment by drewrandall** [comment]
@zvirja maybe this will help illustrate: 

This test:
~~~
[TestClass]
public class AutoFixtureTest
{
    public class TestObject
    {
        public string A { get; set; }
        public string B { get; set; }
        public string C { get; set; }
    }

    [TestMethod]
    public void ShouldPopulateABC()
    {
        var fixture = new Fixture();
        fixture.Customize<TestObject>(x => x.With(p => p.A, "Dog"));
        fixture.Customize<TestObject>(x => x.With(p => p.C, "Cat"));

        var testObject = fixture.Build<TestObject>().With(x => x.A, "Zebra").Create();

        Console.WriteLine($"A: {testObject.A}");
        Console.WriteLine($"B: {testObject.B}");
        Console.WriteLine($"C: {testObject.C}");

        Assert.AreEqual("Zebra", testObject.A, "A is Zebra.");
        Assert.IsNotNull(testObject.B, "B is null.");
        Assert.IsNotNull(testObject.C, "C is null");
    }
}
~~~

Logs this output: 
~~~
A: Zebra
B: B0e6bc76c-c34f-4904-be0d-1deea4a3f1f8
C:
~~~

Why are B and C handled differently?

**Comment by zvirja** [comment]
This is a known buggy behavior. When you configure some property, we add `Omitter` for that property and it prevents automatic property creation later. In your case, we disable automatic creation of the properties `A` and `C`. 

Consider the following known bugs: #321 and #531. The first one seems to be the same as this one, while in that case behavior is an exception.

I could suggest only obvious workarounds here: do not use `Build` after `Customize` or manually assign that object later in the `Build` chain as well.

__P.S.__ Notice, you should chain multiple customizations for the single class, otherwise the latest one will be applied only:

```csharp
fixture.Customize<TestObject>(x => x.With(p => p.B, "Dog").With(p => p.C, "Cat"));
```

**Comment by drewrandall** [comment]
Thanks @zvirja! That clears it up for me.

**Comment by drewrandall** [comment]
@zvirja, considering your tip about chaining the customizations together, how would you take advantage of the ICustomization interface to make the customizations more DRY? Is there a way of using an ICustomization with a ICustomizationComposer? That might be a handy feature.

For example:
```
public class TestObject
{
    public string A { get; set; }
    public string B { get; set; }
    public string C { get; set; }
}

public class AAlwaysEleven : ICustomization
{
    public void Customize(IFixture fixture)
    {
        fixture.Customize<TestObject>(c => c.With(to => to.A, "Eleven"));
    }
}

[TestMethod]
public void SharedCustomizations()
{
    var fixture = new Fixture();
    fixture.Customize(new AAlwaysEleven());
    fixture.Customize<TestObject>(x => x.With(p => p.B, "Twelve"));
    var testObject = fixture.Create<TestObject>();
    Assert.AreEqual("Eleven", testObject.A, "A is Eleven.");
    Assert.AreEqual("Twelve", testObject.B, "A is Twelve.");
}
```

**Comment by zvirja** [comment]
@drewrandall You can easily implement this absent feature in your code, so you will not suffer from the repetitions. Consider the following sample:

Definition:

```csharp
public interface IComposerCustomization<T>
{
    IPostprocessComposer<T> Customize(IPostprocessComposer<T> composer);
}

public static class CustomExtensions
{
    public static IPostprocessComposer<T> With<T>(
        this IPostprocessComposer<T> composer, 
        IComposerCustomization<T> customizer)
    {
        return customizer.Customize(composer);
    }
}
```

Usage:

```csharp
public class TestObject
{
    public string A { get; set; }
    public string B { get; set; }
    public string C { get; set; }
}



public class AAlwaysEleven : ICustomization, IComposerCustomization<TestObject>
{
    public void Customize(IFixture fixture)
    {
        fixture.Customize<TestObject>(composer => this.Customize(composer));
    }

    public IPostprocessComposer<TestObject> Customize(IPostprocessComposer<TestObject> composer)
    {
        return composer.With(x => x.A, "Eleven");
    }
}

[Fact]
public void InnerCustomizationSample()
{
    var fixture = new Fixture();
    fixture.Customize<TestObject>(c => c.With(x => x.B, "Twelve").With(new AAlwaysEleven()));
    
    var testObject = fixture.Create<TestObject>();
    Assert.Equal("Eleven", testObject.A);
    Assert.Equal("Twelve", testObject.B);
}
```

There will be still a small redundancy in the `AAlwaysEleven` like class, but could probably be eliminated even further by extracting a base class for such customizations.

As you can see, AutoFixture is quite friendly library for custom extensions 😉

Also you can check #699 - I provided there a few other ways to solve this issue in even more elegant way 😃

**Comment by drewrandall** [comment]
Thanks @zvirja this looks fantastic! IComposerCustomization seems like a useful addition to AutoFixture to me, but I'll happily embed my own for now.

**Comment by zvirja** [comment]
Fixed in #781. Will be released as a part of v4.
